src/wallet/sqlite.cpp

```diff
@@ -0,0 +1,624 @@
+static constexpr int32_t WALLET_SCHEMA_VERSION = 0;
+
+static Mutex g_sqlite_mutex;
+static int g_sqlite_count GUARDED_BY(g_sqlite_mutex) = 0;
+
+static void ErrorLogCallback(void* arg, int code, const char* msg)
+{
+    // From sqlite3_config() documentation for the SQLITE_CONFIG_LOG option:
+    // "The void pointer that is the second argument to SQLITE_CONFIG_LOG is passed through as
+    // the first parameter to the application-defined logger function whenever that function is
+    // invoked."
+    // Assert that this is the case:
+    assert(arg == nullptr);
+    LogPrintf("SQLite Error. Code: %d. Message: %s\n", code, msg);
+}
+
+SQLiteDatabase::SQLiteDatabase(const fs::path& dir_path, const fs::path& file_path, bool mock)
+    : WalletDatabase(), m_mock(mock), m_dir_path(dir_path.string()), m_file_path(file_path.string())
+{
+    {
+        LOCK(g_sqlite_mutex);
+        LogPrintf("Using SQLite Version %s\n", SQLiteDatabaseVersion());
+        LogPrintf("Using wallet %s\n", m_dir_path);
+
+        if (++g_sqlite_count == 1) {
+            // Setup logging
+            int ret = sqlite3_config(SQLITE_CONFIG_LOG, ErrorLogCallback, nullptr);
+            if (ret != SQLITE_OK) {
+                throw std::runtime_error(strprintf("SQLiteDatabase: Failed to setup error log: %s\n", sqlite3_errstr(ret)));
+            }
+            // Force serialized threading mode
+            ret = sqlite3_config(SQLITE_CONFIG_SERIALIZED);
+            if (ret != SQLITE_OK) {
+                throw std::runtime_error(strprintf("SQLiteDatabase: Failed to configure serialized threading mode: %s\n", sqlite3_errstr(ret)));
+            }
+        }
+        int ret = sqlite3_initialize(); // This is a no-op if sqlite3 is already initialized
+        if (ret != SQLITE_OK) {
+            throw std::runtime_error(strprintf("SQLiteDatabase: Failed to initialize SQLite: %s\n", sqlite3_errstr(ret)));
+        }
+    }
+
+    try {
+        Open(""); // mode is unused
+    } catch (const std::runtime_error&) {
+        // If open fails, cleanup this object and rethrow the exception
+        Cleanup();
+        throw;
+    }
+}
+
+void SQLiteBatch::SetupSQLStatements()
+{
+    int res;
+    if (!m_read_stmt) {
+        if ((res = sqlite3_prepare_v2(m_database.m_db, "SELECT value FROM main WHERE key = ?", -1, &m_read_stmt, nullptr)) != SQLITE_OK) {
+            throw std::runtime_error(strprintf("SQLiteDatabase: Failed to setup SQL statements: %s\n", sqlite3_errstr(res)));
+        }
+    }
+    if (!m_insert_stmt) {
+        if ((res = sqlite3_prepare_v2(m_database.m_db, "INSERT INTO main VALUES(?, ?)", -1, &m_insert_stmt, nullptr)) != SQLITE_OK) {
+            throw std::runtime_error(strprintf("SQLiteDatabase: Failed to setup SQL statements: %s\n", sqlite3_errstr(res)));
+        }
+    }
+    if (!m_overwrite_stmt) {
+        if ((res = sqlite3_prepare_v2(m_database.m_db,"insert or replace into main values(?, ?)", -1, &m_overwrite_stmt, nullptr)) != SQLITE_OK) {
```

Comment:
piconit: a missed space
```suggestion
        if ((res = sqlite3_prepare_v2(m_database.m_db, "insert or replace into main values(?, ?)", -1, &m_overwrite_stmt, nullptr)) != SQLITE_OK) {
```

Reply:
Fixed